File: string/leetcode_344.java
package string;

import java.util.Arrays;

public class leetcode_344 {
    public static void main(String[] args) {
        leetcode_344 lc = new leetcode_344();
        char[] s = {'h', 'e', 'l', 'l', 'o'};
        lc.reverseString(s);
        System.out.println(Arrays.toString(s));
    }

    // it has to be n(1) time complexity.
    public void reverseString(char[] s) {
        int n = s.length;
        for (int i = 0; i < (n / 2); i++) {
            swap(s, i, n - i - 1);
        }
    }

    private void swap(char[] c, int x, int y) {
        char tmp = c[x];
        c[x] = c[y];
        c[y] = tmp;
    }
}

File: string/LeetCode_344.kt
package string

class LeetCode_344 {

    fun reverseString(s: CharArray): Unit {
        val n = s.size

        for (i in 0 until n / 2) {
            swap(s, i, n - i - 1)
        }
    }

    private fun swap(s: CharArray, i: Int, j: Int) {
        val tmp = s[i]
        s[i] = s[j]
        s[j] = tmp
    }
}

fun main() {
    // Input
    val s = charArrayOf('h', 'e', 'l', 'l', 'o')
    // Output
    LeetCode_344().reverseString(s).run {
        println(s.joinToString())
        require(
            charArrayOf('o', 'l', 'l', 'e', 'h').contentEquals(s)
        )
    }
}
